### ISSUE TITLE
Should the json-summary report also have uncovered lines information?

### ISSUE THREAD

**Issue by gazal-k** [istanbuljs/istanbuljs#416]
My teammates and I are trying to report on coverage when running CI. For general coverage information the cobertura format report is quite easy with Jenkins.

For PRs, we'd like to automatically (using a build bot) comment on lines lacking coverage. The json-summary reporter seems quite convenient but it lacks uncovered lines info.

Any thoughts on whether it is ok to modify json-summary to include uncovered lines info?

**Comment by gazal-k** [comment]
@NiQDeveloper, you may wanna watch this issue

**Comment by gazal-k** [comment]
Hey @bcoe, we'd like to contribute to this feature if it is worth having

**Comment by bcoe** [comment]
@gazal-k I haven't personally used the json-summary report, but it seems like having uncovered line information would be worthwhile if it's missing.

Would happily accept a patch.

**Comment by coreyfarrell** [comment]
The json-summary report lists the total and covered line counts.  I'm ok with adding a field for uncovered total, as long as it's added for all types (lines, functions, statements and branches).  I don't think it's appropriate for the json-summary report to give a list of the actual missed lines though.

**Comment by NiQDeveloper** [comment]
Hey @bcoe.

Thanks for the update.

As far as implementation goes - I can see this going in 2 different ways. Either the `FileCoverage.prototype.toSummary` [here](https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-lib-coverage/lib/file.js)  method could be updated to include an uncovered lines array, or the [json-summary](https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-reports/lib/json-summary/index.js) report could be updated to include the `getUncoveredLines` results in `onSummary`.

I'm not sure the downstream impact of using the first method, though it does seem like the cleaner approach to use.

Thoughts?

**Comment by bcoe** [comment]
@NiQDeveloper I don't have a super strong opinion (whatever implementation feels easiest :stuck_out_tongue_closed_eyes:)

To clarify, were you thinking you'd like total uncovered lines (like @coreyfarrell suggests) or were you picturing that the report would include all missed lines -- to @coreyfarrell 's point, it might be better to use another report for this, you could potentially use the Istanbul format JSON itself?

**Comment by NiQDeveloper** [comment]
Hey @bcoe 

I had a look into the JSON format itself, however it is ~10x bigger filesize, and does not summarise the other data.

My reasoning for wanting to include it in the JSON summary, is that it then contains all the same data as the test-summary report.

As far as structure goes, I would target it to look like the below:
```
  "/path/to/file.js": {
    "lines": {
      "total": 8,
      "covered": 10,
      "skipped": 0,
      "pct": 80
    },
    "functions": {
      "total": 3,
      "covered": 4,
      "skipped": 0,
      "pct": 75
    },
    "statements": {
      "total": 4,
      "covered": 5,
      "skipped": 0,
      "pct": 80
    },
    "branches": {
      "total": 8,
      "covered": 8,
      "skipped": 0,
      "pct": 100
    },
    "uncoveredLns": [13, 14, 15]
  }
```

With this data the json-summary file could be used to output the same table that the text-summary outputs in console

**Comment by bcoe** [comment]
@NiQDeveloper to me this seems reasonable. What are your thoughts @coreyfarrell?

**Comment by coreyfarrell** [comment]
@NiQDeveloper I'm confused, `text-summary` does not print any specific missed lines, this is the output of `text-summary` from a project where I added an uncovered function:
```
=============================== Coverage summary ===============================
Statements   : 98.59% ( 70/71 )
Branches     : 100% ( 45/45 )
Functions    : 92.86% ( 13/14 )
Lines        : 98.55% ( 68/69 )
================================================================================
```

The `json-summary` report does currently have the information needed to reproduce the `text-summary` report.  Again it seems to me that this is the point of a summary, that it should give totals, not details.

**Comment by bcoe** [comment]
@coreyfarrell, is there a chance that @NiQDeveloper is referring to the `text` report, which tells folks missing lines/branches?

**Comment by gazal-k** [comment]
Yes, I had the same discussion with @NiQDeveloper and he did mean to compare `json-summary` to the `text` report. They both have coverage reported per source file as opposed to overall coverage as in `text-summary`.

The only thing missing in `json-summary` when compared to `text` is uncovered lines. If we can, I'd add missing coverage per line as a fraction to make it similar to `text`. The text report eventually colorizes that information, doesn't it?
